### PR TITLE
Toward next Major version

### PR DESCRIPTION
**Pull request by nickbalestra**
As per #85 

Goal is to get to a state where we can work together and merge into a single flavor that will allows for both choosing the language (typescript/javascript) and the stream library (xstream, most, rxjs)

Current status

- [x] webpack2
- [x] preset-env
- [x] uglify / minification
- [x] HtmlWebpackPlugin
- [x] JSX -> ProvidePlugin({snabb: 'snabbdom-jsx'})
- [x] inquirer -> for eject confirmation
- [x] inquirer -> for init (choose js or ts, choose streamlib -> rxjs, xstream, most)
- [x] move inquirer to create-cycle-app, in case that no flag --flavor is passed, and pass down answers to init
- [x] adhoc templates for javascript
- [x] adhoc webpack configs for javascript
- [x] Work with yarn if yarn detected
- [x] General cleanup
- [x] Do not use snabbdom-jsx it has bugs, use https://github.com/Swizz/snabbdom-pragma instead

To be done in follow-up PRs:
- adhoc templates for typescript
- adhoc webpack configs for typescript
- cleaner build assets output (js/ css/ images/)
- eslint/prettier
- CSS Autoprefixing
- App Tests (jest/time)
- Move Counter example
- https://github.com/Widdershin/cycle-restart
- Scripts/info based for build

### CODE REVIEW

**Comment by jvanbruegge** [comment]
1. No not use `snabbdom-jsx` it has bugs, thanks to a few PRs from me https://github.com/Swizz/snabbdom-pragma is the way to go
2. Why not use `webpack-blocks`, it makes config a lot easier (for us and for the people after eject)
3. I just merged the `2.0.0` version of the `one-fits-all` flavor, so that might be helpful

**Comment by nickbalestra** [comment]
Thanks!
I prefer to be a bit closer to the "webpack-metal". So that we don't rely on another dependency and users don't have to get another thing. If you eject then you suppose to deal with webpack and is the user issue from that poin onward, so I think is better to leave that decision to him if he want to add or use blocks. 
I find the create-react-app way of commenting the the webpack configs way more of value to an 'eject' user and for us we really know what is going on.

I'm structuring it in a way that Iside `configs` and `template/src` we would have 2 folder, 'javascript' and 'coffescript' so that we can mantain ad-hoc configuration and source file for both. 

hope to push something for you to look at here today/tomorrow

File: packages/cycle-scripts/template/public/index.html [hunk 1]
```diff
@@ -1,15 +1,12 @@
 <!DOCTYPE html>
```
Comment (jvanbruegge) [inline packages/cycle-scripts/template/public/index.html]: This is not needed when you use `index.ejs` and the HTMLWebpackPlugin

File: packages/cycle-scripts/template/src/JavaScript/app.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module.exports = replacements => `${replacements.import}
+
+export function App (sources) {
```
Comment (jvanbruegge) [inline packages/cycle-scripts/template/src/JavaScript/app.js]: I would use the counter example as the base app, like I did [here](https://github.com/cyclejs-community/create-cycle-app-flavors/blob/master/packages/cycle-scripts-one-fits-all/template/src/app.tsx). Has the advantage of showing MVI

File: packages/cycle-scripts/template/src/JavaScript/app.test.js [hunk 3]
```diff
@@ -1,7 +1,8 @@
-// import assert from 'assert'
+module.exports = replacements => `// import assert from 'assert'
```
Comment (jvanbruegge) [inline packages/cycle-scripts/template/src/JavaScript/app.test.js]: Take a look at the new test setup in the [one-fits-all flavor](https://github.com/cyclejs-community/create-cycle-app-flavors/tree/master/packages/cycle-scripts-one-fits-all)

File: packages/cycle-scripts/template/src/JavaScript/app.js (hunk 2 above)
Comment (nickbalestra) [inline packages/cycle-scripts/template/src/JavaScript/app.js]: Totally agree

File: packages/cycle-scripts/template/public/index.html (hunk 1 above)
Comment (nickbalestra) [inline packages/cycle-scripts/template/public/index.html]: no need to use an .ejs with the TMLWebpackPlugin, you can just use an html file and inject into it, no?

Comment (nickbalestra) [inline packages/cycle-scripts/template/public/index.html]: hehe...is . WIP :) not sure i have pushed everything.

Comment (nickbalestra) [inline packages/cycle-scripts/template/public/index.html]: Agree that there should be only 1

**Comment by jvanbruegge** [comment]
What are the dev-utils for?

File: packages/cycle-scripts/template/src/JavaScript/app.test.js (hunk 3 above)
Comment (nickbalestra) [inline packages/cycle-scripts/template/src/JavaScript/app.test.js]: For the tests? Yes those are nice, wi'll defintiely bring/merge all those ins

**Comment by nickbalestra** [comment]
I initially forked them from react-dev-utils, but now using them directly, no need to fork them. It provide a nice little sets of utilities, mainly a great webpackHotDevClient with overlay of errors & co. it make the experience super nice!

**Comment by jvanbruegge** [comment]
https://github.com/facebookincubator/create-react-app/tree/master/packages/react-dev-utils#webpackhotdevclientjs
> It currently supports only Webpack 1.x

**Comment by jvanbruegge** [comment]
We could also incorporate https://github.com/Widdershin/cycle-restart to make hot reloading cooler

**Comment by nickbalestra** [comment]
Lol I guess somehow should do a PR as they are using [webpack2]...

https://pbs.twimg.com/media/B3qIJLFCcAEJLWm.jpg
(https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/package.json#L57) and work fine (I've also tested it without any issue)

yes that will be awesome. Added to the list!

**Comment by nickbalestra** [comment]
@jvanbruegge Once I've cleaned up some stuff, I'll ping you so that you can bring in the typeScripts parts and we can then work together on putting it together. Would be great to launch it at the cycleconf

**Comment by jvanbruegge** [comment]
sure
indeed

**Comment by jvanbruegge** [comment]
Maybe we should also have a third question about test runner, `ava` seems to be quite hot nowadays. So we could provide tests with `mocha-webpack` or `ava-webpack`

**Comment by nickbalestra** [comment]
Imho test-suite choice is a bit out of scope. I'll rather keep the same test suite for the whole monorepo (jest atm). We can always add that later in if we believe is a must.

**Comment by jvanbruegge** [comment]
does jest work with webpack?

**Comment by nickbalestra** [comment]
-> https://facebook.github.io/jest/docs/webpack.html

**Comment by nickbalestra** [comment]
I think I'll brake this PR. Once I got the whole JS part in, we can review it and merge.
Then we can open PRs for 

- moving to counter example
- adding typescript src
- adding tests
- adding css/autoprefixer
- Widdershin/cycle-restart
- eslinting/prettier
- docs

**Comment by jvanbruegge** [comment]
Yeah, i think this is better

**Comment by nickbalestra** [comment]
Cool, 
I'll try to have this one ready for tomorrow so that we can divide and conquer this. Exciting times!
We can put all those releases behind a `next` flag until we hit major

**Comment by nickbalestra** [comment]
@jvanbruegge I've done everything. Few major things:

- CLI if no flavor is passed now prompt user for language (javascript, typescript(not yet working as files are missing)) and stream library (xstream, most, rxjs). 
- If flavor is passed, prompting is delegated to flavor, cycle-scripts support both so that you can run it locally or as a fork with no issue.
- Unique file for configuring strings to interpolate templates and various dependencies (basic, language, and stream-lib related) - `/configs/flavor.js`
- It now support yarn (by autodetecting it (pass `--noyarn` flag if you have yarn but want to force it to use npm instead)
- On bundling, errors are displayed on the page via an overlay
- scripts structure now has `configs/<language>` (typescript to be added), same for `template/src/<language>`
- No need for adding babel config to packaje.json anymore of flavor anymore, directly declared in the webpack.configs as we use webpack node.api)
- and a lot more...

I think this should be a good state to move forward as per our dicussion.
Let me know

**Comment by nickbalestra** [comment]
Addresses #98 #11 #70(via native template literals)

**Comment by jvanbruegge** [comment]
The points are good, but i have few remarks about the code (only had quick look):
- the point about the babelrc was to allow changes, eg for the `env-preset` if you cant di this if its hidden away.
- using the webpack node api means you are leaving traces post-eject (most likely a scripts folder), im not really a fan of this. A partial solution would be to copy all config like normal and have a seperate npm pacjage with the scripts. that way you dont have to have them in your git (dont like this either)
- im not sure if we need seperate js/ts templates. TSC. if you compile ts to es6 you will get your js template for free
i will take a deeper look tomorrow

**Comment by jvanbruegge** [comment]
- There is a lot of duplication in the webpack configs, thats why i like `webpack-blocks`. Even without i would refactor at least to a common file and seperate prod and dev files

**Comment by nickbalestra** [comment]
I know, and you could also use something like [webpack merge](https://github.com/survivejs/webpack-merge). But to be honest, is it that bad to have just two separate plain config files, one for dev and one for prod?

**Comment by jvanbruegge** [comment]
ok

**Comment by nickbalestra** [comment]
Totally agree (also cause it will brake if you select typescript now :D)!

The only version published will be a minor with scoped core-scripts as per https://github.com/cyclejs-community/create-cycle-app/pull/100.

If we need for testing and try I suggest we'll release the work we are now doing behind a npm flag until we get to a rc state

**Comment by jvanbruegge** [comment]
Can you add me to this repo? Then i can assign myself to the tickets and dont have to constantly update my fork

**Comment by nickbalestra** [comment]
Sure, done 😃

**Pull request merged**